### PR TITLE
[NEW FUNCTIONALITY] Add a suspend and resume to the controller.

### PR DESCRIPTION
**Pull request by pwielders**

### CODE REVIEW

File: Source/WPEFramework/PluginServer.cpp [hunk 1]
```diff
@@ -394,9 +394,41 @@ ENUM_CONVERSION_BEGIN(Core::ProcessInfo::scheduler)
...
+            // See if we need can and should SUSPEND.
+            IStateControl* stateControl = _handler->QueryInterface<PluginHost::IStateControl>();
+            if (stateControl == nullptr) {
+                result = Core::ERROR_BAD_REQUEST;
+            }
+            else {
+                // We have a StateControl interface, so at least start suspending, if not already suspended :-)
+                if (stateControl->State() == PluginHost::IStateControl::SUSPENDED) {
+                    result = stateControl->Request(PluginHost::IStateControl::RESUME);
+                    stateControl->Release();
```
Comment (MFransen69) [inline Source/WPEFramework/PluginServer.cpp]: Think this release needs to me moved outside this if condition

File: Source/WPEFramework/PluginServer.cpp [hunk 2]
```diff
@@ -477,6 +509,44 @@ ENUM_CONVERSION_BEGIN(Core::ProcessInfo::scheduler)
...
+                // See if we need can and should SUSPEND.
+                IStateControl* stateControl = _handler->QueryInterface<PluginHost::IStateControl>();
+                if (stateControl == nullptr) {
+                    result = Core::ERROR_BAD_REQUEST;
+                }
+                else {
+                    // We have a StateControl interface, so at least start suspending, if not already suspended :-)
+                    if (stateControl->State() == PluginHost::IStateControl::RESUMED) {
+                        result = stateControl->Request(PluginHost::IStateControl::SUSPEND);
+                        stateControl->Release();
```
Comment (MFransen69) [inline Source/WPEFramework/PluginServer.cpp]: Same as above

File: Source/WPEFramework/json/Controller.json [hunk 3]
```diff
@@ -310,6 +310,95 @@
         }
       ]
     },
+    "Controller.1.resume": {
```
Comment (MFransen69) [inline Source/WPEFramework/json/Controller.json:313]: Think we should also regenerate the documentation from this (but that is not super urgent indeed)

**Review by MFransen69 — CHANGES_REQUESTED** [review]

File: Source/WPEFramework/PluginServer.cpp (hunk 2 above)
Comment (pwielders) [reply Source/WPEFramework/PluginServer.cpp]: good point let me update

File: Source/WPEFramework/PluginServer.cpp [hunk 4]
```diff
@@ -394,9 +394,41 @@ ENUM_CONVERSION_BEGIN(Core::ProcessInfo::scheduler)
...
+        } else if (currentState == IShell::DEACTIVATED) {
+            result = Activate(why);
+        } else if (currentState == IShell::ACTIVATED) {
+            // See if we need can and should SUSPEND.
+            IStateControl* stateControl = _handler->QueryInterface<PluginHost::IStateControl>();
+            if (stateControl == nullptr) {
+                result = Core::ERROR_BAD_REQUEST;
+            }
+            else {
+                // We have a StateControl interface, so at least start suspending, if not already suspended :-)
```
Comment (sebaszm) [inline Source/WPEFramework/PluginServer.cpp]: same here

File: Source/WPEFramework/PluginServer.cpp [hunk 5]
```diff
@@ -394,9 +394,41 @@ ENUM_CONVERSION_BEGIN(Core::ProcessInfo::scheduler)
...
+        IShell::state currentState(State());
+
+        if (currentState == IShell::ACTIVATION) {
+            result = Core::ERROR_INPROGRESS;
+        } else if ((currentState == IShell::DEACTIVATION) || (currentState == IShell::DESTROYED)) {
+            result = Core::ERROR_ILLEGAL_STATE;
+        } else if (currentState == IShell::DEACTIVATED) {
+            result = Activate(why);
+        } else if (currentState == IShell::ACTIVATED) {
+            // See if we need can and should SUSPEND.
```
[the remark this answers is not in the document]
Comment (pwielders) [reply Source/WPEFramework/PluginServer.cpp]: hmmm, seems someone copied :-)

**Review by sebaszm — DISMISSED** [review]

**Review by MFransen69 — DISMISSED** [review]

**Comment by pwielders** [comment]
Although no longer strictly needed, as it is build, maybe good to add..

**Review by MFransen69 — APPROVED** [review]

**Review by sebaszm — APPROVED** [review]

**Pull request merged**